File: tests/go/parse.go
package main

import (
 "os"
 "fmt"
 "time"
 "math"
 "bufio"
 "log"
 "io/ioutil"
 "strings"
 "runtime"
)

func test_full(file_path string) {
  content, err := ioutil.ReadFile(file_path)

  if err != nil {
      log.Fatal(err)
  }

  lines := strings.Split(string(content), "\n")

  for _,line := range lines {
    parse(line)
  }
}

func test_stream(file_path string) {
  file, err := os.Open(file_path)

  if err != nil {
      log.Fatal(err)
  }

  defer file.Close()

  scanner := bufio.NewScanner(file)

  for scanner.Scan() {
    parse(scanner.Text())
  }
}

type Field struct {
  name string
  size int
}

var format = []Field {
  Field{name: "type",               size:  2},
  Field{name: "sequential",         size:  7},
  Field{name: "card_number",        size: 19},
  Field{name: "auth_code",          size:  6},
  Field{name: "sale_date",          size:  8},
  Field{name: "sale_option",        size:  1},
  Field{name: "sale_value",         size: 15},
  Field{name: "installments",       size:  3},
  Field{name: "zero0",              size: 15},
  Field{name: "zero1",              size: 15},
  Field{name: "zero2",              size: 15},
  Field{name: "installment_value",  size: 15},
  Field{name: "resume_number",      size:  7},
  Field{name: "zero3",              size:  3},
  Field{name: "entity_numbeer",     size: 10},
  Field{name: "reserved0",          size: 30},
  Field{name: "status",             size:  2},
  Field{name: "payment_date",       size:  8},
  Field{name: "expirty",            size:  4},
  Field{name: "zero4",              size:  7},
  Field{name: "zero5",              size: 15},
  Field{name: "reserved1",          size:  3},
  Field{name: "error_code",         size:  4},
  Field{name: "ref",                size: 11},
  Field{name: "new_card",           size: 19},
  Field{name: "new_expiry",         size:  4},
  Field{name: "reserved2",          size:  2} }

func parse(line string) map[string]string {
  position := 0
  var parsed = make(map[string]string)

  if len(line) > 0 {
    for _, field := range format {
      parsed[field.name] = line[position:(position+field.size)]
      position = position + field.size
    }
  }

  return parsed
}

func Round(f float64) int {
    if math.Abs(f) < 0.5 {
        return 0
    }
    return int(f + math.Copysign(0.5, f))
}


func print_memory_usage(memory_before, memory_after uint64) {
  memory := ((memory_after - memory_before) / 1024.0)
  memory_usage := Round(float64(memory))

  fmt.Println(fmt.Sprintf("Memory: %d MB", memory_usage))
}

/* This basically parse /proc/PID/status to get RAM usage */
/* In others languages benchmarks, this is achieved by running: ps -o rss= -p */
func capture_memory_usage() uint64 {
  var mem runtime.MemStats
  runtime.ReadMemStats(&mem)

  return (mem.Alloc / 1024)
}

func print_time_spend(time string) {
  fmt.Println(fmt.Sprintf("Time: %s", time))
}

func print_test_type(test_type string, test_mode string, test_name string) {
  str := `
Test Language: GoLang
Test Language Version: go version go1.7.3 linux/amd64
Test Type: %s
Test Name: %s
Test Mode: %s
`

  fmt.Print(fmt.Sprintf(str, test_type, test_name, test_mode))
}

func main() {
  file := os.Getenv("TEST_FILE")
  verbose := os.Getenv("VERBOSE") == "true"
  test_mode := os.Getenv("TEST_MODE")
  test_type := os.Getenv("TEST_TYPE")
  test_name := os.Getenv("TEST_NAME")

  if test_mode == "" { test_mode = "full" }

  if test_type == "" {
    panic("You must suply TEST_TYPE with one the values: read, write, read+write")
  }

  start := time.Now()

  memory_before := capture_memory_usage()

  if verbose == true {
    print_test_type(test_type, test_mode, test_name)
  }

  if test_mode == "full" {
    test_full(file)
  } else if test_mode == "stream" {
    test_stream(file)
  }

  memory_after := capture_memory_usage()

  elapsed := time.Since(start)

  if verbose == true {
    print_time_spend(elapsed.String());
    print_memory_usage(memory_before, memory_after)
  }
}

File: tests/lua/parse.lua
-- please install https://github.com/mah0x211/lua-process
-- this is necessary to use getpid();

process = require('process')

-- http://lua-users.org/wiki/SimpleRound
function round(num, idp)
  local mult = 10^(idp or 0)
  return math.floor(num * mult + 0.5) / mult
end

-- This function used to be a generic function to call any system command
-- but was bugging and now only capture memory usage

function capture_memory_usage(pid)

  local command = ("ps -o rss= -p :pid"):gsub(':pid', pid)

  local io = io.popen(command, 'r')
  local f = assert(io)
  local s = assert(f:read('*a'))
  f:close()
  if raw then return s end
  s = string.gsub(s, '^%s+', '')
  s = string.gsub(s, '%s+$', '')
  s = string.gsub(s, '[\n\r]+', ' ')

  return s
end

function print_memory_usage(memory_before, memory_after)
  local memory_usage = round(((memory_after - memory_before) / 1024.0), 2)

  local memory_str = ("Memory: :memory_usage MB"):gsub(':memory_usage', memory_usage)
  print(memory_str)
end

function print_time_spend(time)
  -- os.clock():
  -- Returns an approximation of the amount in seconds of CPU time used by the program.
  local time_str = ("Time: :time"):gsub(':time', time)
  print(time_str)
end

function test_type_label(test_type, test_name, test_mode)
  local str=[=[
Test Language: Lua
Test Language Version: :version
Test Type: :test_type
Test Name: :test_name
Test Mode: :test_mode]=]
  str = str:gsub(':version', io.popen('lua -v'):read())
  str = str:gsub(':test_type', test_type)
  str = str:gsub(':test_name', test_name)
  str = str:gsub(':test_mode', test_mode)

  return str
end

function test_stream(file_path)
  local file = io.open(file_path)
  local lines = file:lines() -- this returns a file iterator

  -- read line by line
  for line in lines do
    parse(line)
  end

  file:close()
end

function test_full(file_path)
  local file = io.open(file_path)
  local content = file:read("*a")
  file:close()

  for line in string.gmatch(content, "[^%s]+") do
    parse(line)
  end
end

local FORMAT_KEYS = { "type", "sequential", "card_number", "auth_code",
                      "sale_date", "sale_option", "sale_value", "installments",
                      "zero0", "zero1", "zero2", "installment_value", "resume_number",
                      "zero3", "entity_numbeer", "reserved0", "status", "payment_date",
                      "expirty", "zero4", "zero5", "reserved1", "error_code", "ref",
                      "new_card", "new_expiry", "reserved2" }
local FORMAT = {
  ["type"] =               2,
  ["sequential"] =         7,
  ["card_number"] =        19,
  ["auth_code"] =          6,
  ["sale_date"] =          8,
  ["sale_option"] =        1,
  ["sale_value"] =         15,
  ["installments"] =       3,
  ["zero0"] =              15,
  ["zero1"] =              15,
  ["zero2"] =              15,
  ["installment_value"] =  15,
  ["resume_number"] =      7,
  ["zero3"] =              3,
  ["entity_numbeer"] =     10,
  ["reserved0"] =          30,
  ["status"] =             2,
  ["payment_date"] =       8,
  ["expirty"] =            4,
  ["zero4"] =              7,
  ["zero5"] =              15,
  ["reserved1"] =          3,
  ["error_code"] =         4,
  ["ref"] =                11,
  ["new_card"] =           19,
  ["new_expiry"] =         4,
  ["reserved2"] =          2
}
function parse(line)
  local position = 0
  parsed = {}
  for i, key in ipairs(FORMAT_KEYS) do
    local size = FORMAT[key]
    parsed[key] = string.sub(line, position+1, position + size)
    position = position + size
  end
  return parsed
end

function basename(str)
  value = str:gsub("(.*/)(.*)", '%2')

  return value
end

-- get options from system env(MODE=stream VERBOSE=true lua lua/read.lua)
local file_path = os.getenv('TEST_FILE')
local test_type = os.getenv('TEST_TYPE')
local test_mode = (os.getenv('TEST_MODE') or 'full')

local test_name = (os.getenv('TEST_NAME') or basename(file_path))
local verbose   = (os.getenv('VERBOSE') == 'true')

function run(file_path, test_mode, test_type, test_name, verbose)
  if verbose then
    print(test_type_label(test_type, test_name, test_mode))
  end

  local memory_before = capture_memory_usage(process.getpid())

  if(test_mode == 'full') then
    test_full(file_path)
  else
    test_stream(file_path)
  end

  local memory_after = capture_memory_usage(process.getpid())

  if verbose then
    print_memory_usage(memory_before, memory_after)
    print_time_spend(os.clock())
  end
end

run(file_path, test_mode, test_type, test_name, verbose)

--[===[]
process = require('process')

IOBench = {}
IOBench.__index = IOBench

-- Extend math
-- http://lua-users.org/wiki/SimpleRound
function math.round(num, idp)
  local mult = 10^(idp or 0)
  return math.floor(num * mult + 0.5) / mult
end

function IOBench.new(file_path, test_mode, verbose)
  local self = setmetatable({}, IOBench)
  self.file_path = init
  self.mode = test_mode
  self.verbose = verbose
  return self
end

function IOBench.run()
end

function IOBench.print_memory_usage(memory_before, memory_after)
  memory_usage = math.round(((memory_after - memory_before) / 1024.0), 2)

  print(("Memory: :memory_usage MB"):gsub(':memory_usage', memory_usage))
end
--- ]===]
